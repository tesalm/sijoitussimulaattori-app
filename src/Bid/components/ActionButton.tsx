import React from 'react';

import { Text, View } from 'react-native';
import { WizardFormColors } from '../../App/colors';
import { t } from '../../assets/i18n';
import Icon from '../../general/icon';
import { actionButtons } from '../styles';

export interface ActionButtonProps {
  action: string;
  active: boolean;
}

export const ActionButton = (props: ActionButtonProps): JSX.Element => {
  return (
    <View style={actionButtons.buttonWithText}>
      <Icon
        iconName={props.action}
        iconHeight={24}
        iconWidth={24}
        tintColor={
          props.active
            ? WizardFormColors.buttonsActive
            : WizardFormColors.buttonsUnactive
        }
      />
      <Text
        style={
          props.active
            ? actionButtons.buttonActive
            : actionButtons.buttonUnactive
        }
      >
        {props.action === 'buy' && t('BidPage.Buy')}
        {props.action === 'sell' && t('BidPage.Sell')}
      </Text>
    </View>
  );
};
