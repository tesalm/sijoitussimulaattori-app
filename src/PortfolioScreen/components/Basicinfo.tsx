import React from 'react';
import { ActivityIndicator, Text, View } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { t } from '../../assets/i18n';
import { RootState } from '../../redux/reducers';
import { Portfolio } from '../reducers';
import { stockStyles } from '../styles';

export interface BasicinfoProps {
  portfolio?: Portfolio;
  loading?: boolean;
  error?: Error;
}

export class Basicinfo extends React.Component<BasicinfoProps> {
  constructor(props: BasicinfoProps) {
    super(props);
  }
  revenueColor = (revenue: number): typeof stockStyles.revenueValueGreen => {
    return revenue >= 0
      ? stockStyles.revenueValueGreen
      : stockStyles.revenueValueRed;
  };

  formatValue = (value: number, currency: string): string => {
    if (currency == 'USD') {
      return value + ' $';
    } else if (currency == 'EUR') {
      return value + ' €';
    }
    return value + ' $';
  };

  render() {
    const { portfolio, loading, error } = this.props;
    if (loading == undefined && portfolio == undefined && error == undefined) {
      return (
        <View>
          <Text>vituiks meni</Text>
        </View>
      );
    }

    if (loading) {
      return (
        <View style={stockStyles.loading}>
          <ActivityIndicator size="large" />
        </View>
      );
    } else {
      if (portfolio == undefined) {
        //TODO: Format the error message to user
        if (error) {
          return <Text>Error! {error.message} </Text>;
        } else {
          return <Text>Error! asöfeh {portfolio} </Text>;
        }
      }
    }

    return (
      <View>
        <View style={stockStyles.basicinfo}>
          <Text style={stockStyles.titleStyle}>{t('PortfolioPage.Title')}</Text>
        </View>
        <View style={stockStyles.basicinfo}>
          <View style={stockStyles.basicinfoSmallerComp}>
            <Text style={stockStyles.valueHeader}>
              {t('PortfolioPage.MarketValue')}
            </Text>
            <Text style={stockStyles.value}>{portfolio.marketValue}</Text>
            <Text style={stockStyles.valueHeader}>
              {t('PortfolioPage.Cash')}
            </Text>
            <Text style={stockStyles.value}>{portfolio.cash} </Text>
          </View>
          <View style={stockStyles.basicinfoSmallerComp}>
            <Text style={stockStyles.valueHeaderMiddle}>
              {t('PortfolioPage.TotalValue')}
            </Text>
            <Text style={stockStyles.valueMiddle}>
              {portfolio.marketValue + portfolio.cash}
            </Text>
          </View>

          <View style={stockStyles.basicinfoMidComp}>
            <Text style={stockStyles.valueHeaderRightSide}>
              {t('PortfolioPage.Revenue')}
            </Text>
            <Text style={this.revenueColor(4.52)}>{portfolio.revenue}</Text>
          </View>
        </View>
      </View>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  portfolio: state.singlePortfolio.portfolio,
  loading: state.singlePortfolio.loading,
  error: state.singlePortfolio.error,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators({}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Basicinfo);
