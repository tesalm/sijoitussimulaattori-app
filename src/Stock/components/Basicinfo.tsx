import React from 'react';

import { Text, View, ActivityIndicator } from 'react-native';

import { t } from '../../assets/i18n';
import { stockStyles } from '../styles';

import { Intraday, Metadata } from '../reducers';
import { RootState } from '../../redux/reducers';
import { Dispatch, bindActionCreators } from 'redux';
import { connect } from 'react-redux';

export interface BasicinfoProps {
  metadata?: Metadata;
  intraday?: Intraday;
  metaLoading?: boolean;
  metaError?: Error;
  intraLoading?: boolean;
  intraError?: Error;
  revenue: string;
}

export class Basicinfo extends React.Component<BasicinfoProps> {
  constructor(props: BasicinfoProps) {
    super(props);
  }

  revenueColor = (revenue: number): typeof stockStyles.revenueValueGreen => {
    return revenue >= 0
      ? stockStyles.revenueValueGreen
      : stockStyles.revenueValueRed;
  };

  formatValue = (value: number, currency: string): string => {
    if (currency == 'USD') {
      return value + ' $';
    } else if (currency == 'EUR') {
      return value + ' €';
    }
    return value + ' $';
  };

  render() {
    const {
      metadata,
      intraday,
      metaLoading,
      metaError,
      intraLoading,
      intraError,
      revenue,
    } = this.props;

    if (metaLoading || intraLoading) {
      return (
        <View style={stockStyles.loading}>
          <ActivityIndicator size="large" />
        </View>
      );
    } else {
      if (intraday === undefined || metadata === undefined) {
        let errorMessage;
        if (metaError) {
          errorMessage = metaError.message + ' ';
        }
        if (intraError) {
          errorMessage = errorMessage + intraError.message;
        }
        // TODO: Muokkaa error-teksti käyttäjälle.
        return <Text>Error! {errorMessage} </Text>;
      }
    }

    return (
      <View>
        <View style={stockStyles.basicinfo}>
          <Text style={stockStyles.titleStyle}>
            {metadata.name} ({metadata.symbol})
          </Text>
        </View>
        <View style={stockStyles.basicinfo}>
          <View style={stockStyles.basicinfoSmallerComp}>
            <Text style={stockStyles.valueHeader}>{t('StockPage.Low')}</Text>
            <Text style={stockStyles.value}>
              {this.formatValue(intraday.low, metadata.currency)}
            </Text>
            <Text style={stockStyles.valueHeader}>{t('StockPage.High')}</Text>
            <Text style={stockStyles.value}>
              {this.formatValue(intraday.high, metadata.currency)}
            </Text>
          </View>
          <View style={stockStyles.basicinfoSmallerComp}>
            <Text style={stockStyles.valueHeader}>{t('StockPage.Open')}</Text>
            <Text style={stockStyles.value}>
              {this.formatValue(intraday.open, metadata.currency)}
            </Text>
            <Text style={stockStyles.valueHeader}>{t('StockPage.Close')}</Text>
            <Text style={stockStyles.value}>
              {this.formatValue(intraday.close, metadata.currency)}
            </Text>
          </View>
          <View style={stockStyles.basicinfoSmallerComp} />
          <View style={stockStyles.basicinfoMidComp}>
            <Text style={stockStyles.valueHeaderRightSide}>
              {t('StockPage.Volume')}
            </Text>
            <Text style={stockStyles.valueRightSide}>{intraday.volume}</Text>
            <Text style={stockStyles.valueHeaderRightSide}>
              {t('StockPage.RevenueText')}
            </Text>
            <Text style={this.revenueColor(intraday.open)}>{revenue}</Text>
          </View>
        </View>
        <View style={stockStyles.basicinfo}>
          <View>
            <Text style={stockStyles.valueHeader}>
              {t('StockPage.Updated')}: {intraday.fetchTime.toLocaleString()}
            </Text>
          </View>
        </View>
      </View>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  metadata: state.singleStock.metadata,
  intraday: state.singleStock.intraday,
  metaLoading: state.singleStock.metaLoading,
  metaError: state.singleStock.metaError,
  intraLoading: state.singleStock.intraLoading,
  intraError: state.singleStock.intraError,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators({}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Basicinfo);
