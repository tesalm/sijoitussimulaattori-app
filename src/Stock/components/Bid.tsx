import React from 'react';
import { Text, View, TouchableOpacity } from 'react-native';
import { stockStyles } from '../styles';
import Icon from '../../general/icon';

import { t } from '../../assets/i18n';
import { NavigationScreenProps } from 'react-navigation';
import { Colors } from '../../App/colors';

interface BidProps {}

type BidPropsWithNavigation = BidProps & NavigationScreenProps;

const Bid = (props: BidPropsWithNavigation): JSX.Element => {
  return (
    <TouchableOpacity onPress={() => props.navigation.navigate('Bid')}>
      <View style={stockStyles.bidView}>
        <View style={stockStyles.bidLogoView}>
          <Icon
            iconName={'transaction'}
            iconHeight={24}
            iconWidth={24}
            tintColor={Colors.baseColor}
          />
          <Text style={stockStyles.bidText}>{t('StockPage.Bid')}</Text>
        </View>
        <Icon
          iconName={'open'}
          iconHeight={24}
          iconWidth={24}
          tintColor={Colors.baseColor}
        />
      </View>
    </TouchableOpacity>
  );
};

export default Bid;
