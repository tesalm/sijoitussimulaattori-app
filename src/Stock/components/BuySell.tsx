import React from 'react';
import { Text, TouchableHighlight, View } from 'react-native';
import { stockStyles } from '../styles';
import Icon from '../../general/icon';

import { t } from '../../assets/i18n';

interface BuySellProps {
  containerName: string;
}

const BuySell = (props: BuySellProps): JSX.Element => {
  return (
    // TODO: Make onPress for these items that go to buy or sell view.
    <TouchableHighlight style={stockStyles.buySellButton}>
      <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
        <Text style={stockStyles.buySellText}>{t(props.containerName)}</Text>
        <Icon iconName={'close.png'} iconHeight={24} iconWidth={24} />
      </View>
    </TouchableHighlight>
  );
};

export default BuySell;
