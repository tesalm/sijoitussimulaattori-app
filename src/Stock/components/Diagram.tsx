import React from 'react';
import { Text, View, ActivityIndicator } from 'react-native';

import { t } from '../../assets/i18n';
import { stockStyles } from '../styles';
import { HistoryData } from '../../MarketScreen/reducers';

interface DiagramProps {
  historyData?: HistoryData;
  historyLoading?: boolean;
  historyError?: Error;
}

const Diagram = (props: DiagramProps): JSX.Element => {
  if (props.historyLoading) {
    return (
      <View style={stockStyles.loading}>
        <ActivityIndicator size="large" />
      </View>
    );
  } else {
    if (props.historyError) {
      // TODO: Muokkaa error-teksti käyttäjälle.
      return <Text>Error! {props.historyError.message}</Text>;
    }
  }

  return (
    <View>
      <Text style={stockStyles.titleStyle}>
        {t('StockPage.RevenueOverYear')}
      </Text>
    </View>
  );
};
export default Diagram;
