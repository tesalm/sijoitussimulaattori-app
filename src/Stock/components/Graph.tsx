import React from 'react';
import { Text } from 'react-native';

import { t } from '../../assets/i18n';
import { HistoryDataQuote } from '../../MarketScreen/reducer';
import { verticalScale } from '../../util/scale';
import { stockContainerStyles } from '../styles';

const { VictoryArea, VictoryChart, VictoryAxis } = require('victory-native');

interface graphProps {
  historydata: Array<HistoryDataQuote>;
}

const Graph = (props: graphProps): JSX.Element => {
  const { historydata } = props;
  if (historydata.length < 2) {
    return <Text>{t('Graph.NoData')}</Text>;
  }
  const maxDate = new Date(historydata[0].date);
  const minDate = new Date(historydata[historydata.length - 1].date);
  const startDate = new Date(
    minDate.getFullYear(),
    minDate.getMonth(),
    minDate.getDate()
  ).getTime();
  const endDate = new Date(
    maxDate.getFullYear(),
    maxDate.getMonth(),
    maxDate.getDate()
  ).getTime();
  const months = ['1','2','3','4','5','6','7','8','9','10','11','12']
          .map((key) => t('Graph.Month.' + key));

  function getDomain(data: Array<HistoryDataQuote>) {
    let min = data[0].close;
    let max = data[0].close;
    for (let i = 1, len = data.length; i < len; i++) {
      let v = data[i].close;
      min = v < min ? v : min;
      max = v > max ? v : max;
    }
    return [min, max];
  }

  return (
    <VictoryChart
      height={verticalScale(180)}
      animate={false}
      padding={stockContainerStyles.graphPadding}
      domainPadding={{ y: verticalScale(20) }}
      scale={{ x: 'time' }}
    >
      {/* x -axis */}
      <VictoryAxis
        invertAxis={true}
        fixLabelOverlap={true}
        tickFormat={(x: string) => {
          if (historydata[historydata.length - 1].date === x) {
            return null;
          }
          const date = new Date(x);
          if (endDate === startDate) {
            let mm = date.getMinutes().toString();
            if (mm.length === 1) {
              mm = '0' + mm;
            }
            return date.getUTCHours() + ':' + mm + '\t';
          }
          if (minDate.getFullYear() !== maxDate.getFullYear()) {
            const yy = date
              .getFullYear()
              .toString()
              .substr(-2);
            return date.getDate() + '. ' + months[date.getMonth()] + ' - ' + yy;
          }
          return date.getDate() + '. ' + months[date.getMonth()] + '\t';
        }}
        style={{
          tickLabels: {
            fontSize: 10,
            fill: '#455A64',
            padding: 2,
            textAnchor: 'end',
          },
          axis: { stroke: 'none' },
        }}
      />

      {/* y -axis */}
      <VictoryAxis
        dependentAxis
        orientation="right"
        tickFormat={(y: number) => {
          if (y >= 1000) {
            return Math.round(y);
          }
          return y;
        }}
        style={{
          tickLabels: {
            fontSize: 10,
            fill: '#455A64',
            padding: 2,
          },
          axis: { stroke: 'none' },
          grid: { stroke: 'gray', opacity: 0.1 },
        }}
      />

      <VictoryArea
        domain={{ y: getDomain(historydata) }}
        interpolation="linear"
        animate={false}
        style={{
          data: {
            fill: '#FF9900', //#00A8EF light blue
            stroke: '#FF8000',
            strokeWidth: 1,
            fillOpacity: 0.3,
          },
        }}
        data={historydata}
        x="date"
        y="close"
      />
    </VictoryChart>
  );
};

export default Graph;
