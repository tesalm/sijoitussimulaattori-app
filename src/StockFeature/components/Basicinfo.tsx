import React from 'react';
import { Text, View, StyleSheet, TextStyle } from 'react-native';

import { t } from '../../assets/i18n';
import { stockStyles } from '../styles';

const Basicinfo = (): JSX.Element => {
  return (
    <View style={{ flex: 3, justifyContent: 'space-between' }}>
      <View style={{ flex: 2, flexDirection: 'row'  }}>
        <View style={{ width: '25%', flexDirection: 'column' }}>
          <Text style={ stockStyles.infoHeader }>{t('Bid')}</Text>
          <Text style={ stockStyles.infoText }>10€</Text>
          <Text style={ stockStyles.infoHeader }>{t('Offer')}</Text>
          <Text style={ stockStyles.infoText }>500€</Text>
        </View>
        <View style={{ width: '25%', flexDirection: 'column' }}>
          <Text style={ stockStyles.infoHeader }>{t('High')}</Text>
          <Text style={ stockStyles.infoText }>10€</Text>
          <Text style={ stockStyles.infoHeader }>{t('Low')}</Text>
          <Text style={ stockStyles.infoText }>500€</Text>
        </View>
        <View style={{ width: '50%', flexDirection: 'column'}}>
          <Text>{t('Market value')}</Text>
          <Text>8</Text>
          <Text>{t('Revenue in 24h')}</Text>
          <Text>7</Text>
        </View>
      </View>
      <View style={{ flex: 1 }}>
        <Text style={ stockStyles.infoHeader }>{t('Updated')} 09.11.2018 16:36.48</Text>
      </View>
    </View>
  );
};

export default Basicinfo;
