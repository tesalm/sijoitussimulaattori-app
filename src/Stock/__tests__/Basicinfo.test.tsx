import * as React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import { BasicinfoProps, Basicinfo } from '../components/Basicinfo';

describe('basicinfo tests', () => {
  // Mock for navigation.
  const navigationMock: any = {};
  const middlewares: any = [];
  const mockStore = configureStore(middlewares);

  const defaultBasicinfoProps: BasicinfoProps = {
    metadata: undefined,
    intraday: undefined,
    metaLoading: false,
    intraLoading: false,
    metaError: undefined,
    intraError: undefined,
    revenue: '',
  };

  const errorBasicinfoProps: BasicinfoProps = {
    metadata: undefined,
    intraday: undefined,
    metaLoading: false,
    intraLoading: false,
    metaError: { name: 'Metadata error', message: 'Metadata not found' },
    intraError: undefined,
    revenue: '',
  };

  const successBasicinfoProps: BasicinfoProps = {
    metadata: {
      symbol: 'AAPL',
      name: 'Apple Inc.',
      type: 'string',
      region: 'string',
      marketOpen: '2018-12-12 11:20:00',
      marketClose: '2018-12-12 11:20:00',
      timeZone: 'US/Eastern',
      currency: 'USD',
      fetchTime: new Date(),
    },
    intraday: {
      symbol: 'AAPL',
      date: '2018-12-12',
      open: 170.4,
      high: 170.43,
      low: 165.43,
      close: 168.63,
      volume: 12279994,
      fetchTime: new Date(),
    },
    metaLoading: false,
    intraLoading: false,
    metaError: undefined,
    intraError: undefined,
    revenue: '',
  };

  it('Basicinfo renders correctly', async () => {
    const wrapper = shallow(
      <Basicinfo {...defaultBasicinfoProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('Basicinfo renders correctly with error', async () => {
    const wrapper = shallow(
      <Basicinfo {...errorBasicinfoProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('Basicinfo renders correctly with stock', async () => {
    const wrapper = shallow(
      <Basicinfo {...successBasicinfoProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
