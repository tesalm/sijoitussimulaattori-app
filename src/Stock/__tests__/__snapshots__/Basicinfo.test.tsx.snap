// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basicinfo tests Basicinfo renders correctly 1`] = `
<TouchableText
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  forwardedRef={null}
>
  Error! 
   
</TouchableText>
`;

exports[`basicinfo tests Basicinfo renders correctly with error 1`] = `
<TouchableText
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  forwardedRef={null}
>
  Error! 
  Metadata not found 
   
</TouchableText>
`;

exports[`basicinfo tests Basicinfo renders correctly with stock 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
